### ISSUE TITLE
ngx.flush not blocking (HTTP2 only)

### ISSUE THREAD

**Issue by juancampa** [openresty/lua-nginx-module#1663]
## TL;DR
Streaming doesn't work correctly when responding to HTTP V2 requests because `ngx.flush` never blocks.
```
  -- Stream out the response
  ngx.print(chunk)
  ngx.flush(true)
```

## Repro

Easy repro steps:
```
$ mkdir certs conf.d

Generate self-signed cert to serve HTTP V2
$ openssl req -x509 -nodes -days 365 -newkey rsa:2048 -keyout certs/nginx-selfsigned.key -out certs/nginx-selfsigned.crt

Download minimal repro case
$ wget https://gist.githubusercontent.com/juancampa/634f9739a108bc5ea154c34e19343314/raw/ -O conf.d/default.conf

Run vanilla OpenResty build
$ docker run -v $PWD/certs:/etc/certs -v $PWD/conf.d:/etc/nginx/conf.d -p 3000:443 openresty/openresty:alpine

Repro it!
$ curl https://localhost:3004/test --http2 --limit-rate 600m -o out.txt --insecure
```

Result:
```
curl: (18) transfer closed with 86786048 bytes remaining to read
```

stderr:
```
STILL GOING 0.078125
STILL GOING 0.15625
STILL GOING 0.234375
STILL GOING 0.3125
STILL GOING 0.390625
STILL GOING 0.46875
STILL GOING 0.546875
STILL GOING 0.625
STILL GOING 0.703125
STILL GOING 0.78125
ERROR PRINTING: nginx output filter error <--- Problem
```

## Details

The problem only happens if both of these things are true:

 - `http2` is being used
 - The transfer speed is slow compared to the speed of data being `ngx.print`ed

## Root Cause

I'm still investigating but it seems like the issue is related to [this condition](https://github.com/openresty/lua-nginx-module/blob/4c4bb0fcbd12fc79e5adc5fad5d454dc03be875f/src/ngx_http_lua_output.c#L551) never being true for HTTP2.

The problem goes away if I also check for `NGX_HTTP_V2_BUFFERED` being set. But I'm not sure that's the way to go. My guess is that we're not setting `NGX_HTTP_LOWLEVEL_BUFFERED` somewhere.

Thanks for your help!

**Comment by agentzh** [comment]
Thanks for the report. We'll look into this.

**Comment by zhuizhuhaomeng** [comment]
``` gdb
(gdb) bt
#0  ngx_http_v2_filter_get_data_frame (last=0x7fae62263fb0, first=0x7fae62263fb0, len=8192, stream=0x7fae622e9f28) at src/http/v2/ngx_http_v2_filter_module.c:1698
#1  ngx_http_v2_send_chain (fc=0x7fae6ab98100, in=<optimized out>, limit=988766208) at src/http/v2/ngx_http_v2_filter_module.c:1581
#2  ngx_http_write_filter (r=<optimized out>, in=<optimized out>) at src/http/ngx_http_write_filter_module.c:294
#3  ngx_http_chunked_body_filter (r=0x7fae622e9050, in=0x7fae62263f30) at src/http/modules/ngx_http_chunked_filter_module.c:115
#4  ngx_http_gzip_body_filter (r=0x7fae622e9050, in=0x7fae62263f30) at src/http/modules/ngx_http_gzip_filter_module.c:308
#5  ngx_http_ssi_body_filter (r=0x7fae622e9050, in=<optimized out>) at src/http/modules/ngx_http_ssi_filter_module.c:449
#6  ngx_http_charset_body_filter (r=0x7fae622e9050, in=0x7fae62263f30) at src/http/modules/ngx_http_charset_filter_module.c:647
#7  ngx_http_addition_body_filter (in=<optimized out>, r=<optimized out>) at src/http/modules/ngx_http_addition_filter_module.c:149
#8  ngx_http_addition_body_filter (r=0x7fae622e9050, in=0x7fae62263f30) at src/http/modules/ngx_http_addition_filter_module.c:133
#9  ngx_http_gunzip_body_filter (r=0x7fae622e9050, in=0x7fae62263f30) at src/http/modules/ngx_http_gunzip_filter_module.c:185
#10 ngx_http_trailers_filter (r=0x7fae622e9050, in=0x7fae62263f30) at src/http/modules/ngx_http_headers_filter_module.c:274
#11 ngx_http_lua_capture_body_filter (r=0x7fae622e9050, in=0x7fae62263f30) at ../ngx_lua-0.10.18rc3/src/ngx_http_lua_capturefilter.c:133
#12 ngx_output_chain (ctx=ctx@entry=0x7fae6771cbc8, in=in@entry=0x7fae62263f30) at src/core/ngx_output_chain.c:74
#13 ngx_http_copy_filter (r=0x7fae622e9050, in=0x7fae62263f30) at src/http/ngx_http_copy_filter_module.c:152
#14 ngx_http_output_filter (r=0x7fae622e9050, in=0x7fae62263f30) at src/http/ngx_http_core_module.c:1856
#15 ngx_http_lua_output_filter (r=r@entry=0x7fae622e9050, in=<optimized out>, in@entry=0x7fae62263f30) at ../ngx_lua-0.10.18rc3/src/ngx_http_lua_util.c:676
#16 ngx_http_lua_send_chain_link (r=r@entry=0x7fae622e9050, ctx=ctx@entry=0x7fae6771c7e0, in=in@entry=0x7fae62263f30) at ../ngx_lua-0.10.18rc3/src/ngx_http_lua_util.c:630
#17 ngx_http_lua_send_chain_link (r=r@entry=0x7fae622e9050, ctx=ctx@entry=0x7fae6771c7e0, in=in@entry=0x7fae62263f30) at ../ngx_lua-0.10.18rc3/src/ngx_http_lua_util.c:534
#18 ngx_http_lua_ngx_echo (L=0x7fae6d7822c0, newline=<optimized out>) at ../ngx_lua-0.10.18rc3/src/ngx_http_lua_output.c:240
#19 lj_BC_FUNCC () from target:/usr/local/openresty/luajit/lib/libluajit-5.1.so.2 #20  ngx_http_lua_run_thread (L=L@entry=0x7fae6d7ad380, r=r@entry=0x7fae622e9050, ctx=ctx@entry=0x7fae6771c7e0, nrets=<optimized out>, nrets@entry=0) at ../ngx_lua-0.10.18rc3/src/ngx_http_lua_util.c:1132
#21 ngx_http_lua_access_by_chunk (L=0x7fae6d7ad380, r=0x7fae622e9050) at ../ngx_lua-0.10.18rc3/src/ngx_http_lua_accessby.c:337
#22 ngx_http_lua_access_handler (r=0x7fae622e9050) at ../ngx_lua-0.10.18rc3/src/ngx_http_lua_accessby.c:158
#23 ngx_http_core_access_phase (r=0x7fae622e9050, ph=0x7fae6acd53b8) at src/http/ngx_http_core_module.c:1103
#24 ngx_http_core_run_phases (r=r@entry=0x7fae622e9050) at src/http/ngx_http_core_module.c:878
#25 ngx_http_handler (r=r@entry=0x7fae622e9050) at src/http/ngx_http_core_module.c:861 
#26 ngx_http_process_request (r=r@entry=0x7fae622e9050) at src/http/ngx_http_request.c:2081
#27 ngx_http_v2_run_request (r=0x7fae622e9050) at src/http/v2/ngx_http_v2.c:3928 
#28 ngx_http_v2_state_header_complete (end=0x7fae676a03b1 "", pos=0x7fae676a03a8 "", h2c=0x7fae6aceb000) at src/http/v2/ngx_http_v2.c:1865
#29 ngx_http_v2_state_header_complete (h2c=0x7fae6aceb000, pos=0x7fae676a03a8 "", end=0x7fae676a03b1 "") at src/http/v2/ngx_http_v2.c:1842
#30 ngx_http_v2_state_field_len (h2c=h2c@entry=0x7fae6aceb000, pos=<optimized out>, end=end@entry=0x7fae676a03b1 "") at src/http/v2/ngx_http_v2.c:1528
#31 ngx_http_v2_state_header_block (h2c=0x7fae6aceb000, pos=<optimized out>, end=0x7fae676a03b1 "") at src/http/v2/ngx_http_v2.c:1441
#32 ngx_http_v2_read_handler (rev=0x7fae6ad37f60) at src/http/v2/ngx_http_v2.c:422
#33 ngx_epoll_process_events (cycle=0x7fae6abf8850, timer=<optimized out>, flags=<optimized out>) at src/event/modules/ngx_epoll_module.c:901
#34 ngx_process_events_and_timers (cycle=cycle@entry=0x7fae6abf8850) at src/event/ngx_event.c:257
#35 ngx_worker_process_cycle (cycle=0x7fae6abf8850, data=<optimized out>) at src/os/unix/ngx_process_cycle.c:811
#36 ngx_spawn_process (cycle=cycle@entry=0x7fae6abf8850, proc=proc@entry=0x44dcf0 <ngx_worker_process_cycle>, data=data@entry=0x0, name=name@entry=0x570a38 "worker process", respawn=respawn@entry=-3) at src/os/unix/ngx_process.c:199
#37 ngx_start_worker_processes (cycle=cycle@entry=0x7fae6abf8850, n=1, type=type@entry=-3) at src/os/unix/ngx_process_cycle.c:387
#38 ngx_master_process_cycle (cycle=0x7fae6abf8850) at src/os/unix/ngx_process_cycle.c:135 
#39 main (argc=<optimized out>, argv=<optimized out>) at src/core/nginx.c:385
```

For now, it returns an error because nginx found there are too many frames in connection(over 10000).
The root cause is just as juancampa's judgment. ngx_http_lua_ngx_flush function never wait for flush.

**Comment by zhuizhuhaomeng** [comment]
https://github.com/openresty/lua-nginx-module/pull/1812
